File: api/app/domain/category.py
from typing import Optional
from uuid import UUID

from app.domain.core import DomainModel, InputModel
from app.domain.user import SYSTEM_USER

CATEGORY_TABLE = "categories"
DEFAULT_CATEGORY_ID = UUID("6c47e0cc-b47c-4661-bda3-8e8077fed6c7")


class Category(DomainModel):
    name: str
    parentCategory: Optional[UUID] = None
    owner: UUID


class CategoryInput(InputModel):
    name: str
    parentCategory: Optional[UUID] = None
    # TODO: Remove this once it can be discerned by the API from the user:
    owner: UUID


class Subcategory(DomainModel):
    name: str
    parentCategory: UUID
    owner: UUID


class PrimaryCategory(DomainModel):
    name: str
    owner: UUID
    subcategories: list[Subcategory]


def decode_category_input_model(uuid: UUID, input: CategoryInput) -> Category:
    return Category(
        id=uuid,
        name=input.name,
        parentCategory=input.parentCategory,
        owner=input.owner,
    )


UNCATEGORIZED = Category(
    id=UUID("6c47e0cc-b47c-4661-bda3-8e8077fed6c7"),
    name="Uncategorized",
    owner=SYSTEM_USER.id,
)
INCOME = Category(
    id=UUID("a3720dcc-0ba4-426d-9c41-620a0fbe0ad6"),
    name="Income",
    owner=SYSTEM_USER.id,
)
REFUND = Category(
    id=UUID("7251dda2-7f20-4fd6-9e23-3d5e1b0145ce"),
    name="Return / Reimbursement / Refund",
    owner=SYSTEM_USER.id,
)
TRANSFER = Category(
    id=UUID("c37846fa-1ee0-4e1d-95dd-5c46e13e08c8"),
    name="Transfer",
    owner=SYSTEM_USER.id,
)

SYSTEM_CATEGORIES = [
    UNCATEGORIZED,
    INCOME,
    REFUND,
    TRANSFER,
]

File: client/src/schemas/category.ts
import * as yup from "yup";
import { Labels, SYSTEM_USER, pagedResultsSchema } from "./shared";

export const INCOME = "a3720dcc-0ba4-426d-9c41-620a0fbe0ad6";
export const UNCATEGORIZED = "6c47e0cc-b47c-4661-bda3-8e8077fed6c7";

export const categorySchema = yup.object({
  id: yup.string().required(),
  name: yup.string().required(),
  parentCategory: yup.string().nullable(),
  owner: yup.string().required(),
});

export const categorySchemaList = yup.array().of(categorySchema).required();

export const pagedCategoriesSchema = pagedResultsSchema.shape({
  results: categorySchemaList,
});

// TODO: Add .omit(["owner"]) once user login is set up.
export const categoryInputSchema = categorySchema.shape({
  id: yup.string().nullable(),
});

export const subcategorySchema = categorySchema.shape({
  parentCategory: yup.string().required(),
});

export const primaryCategorySchema = categorySchema
  .shape({
    subcategories: yup.array().of(categorySchema).required(),
  })
  .omit(["parentCategory"]);

export const primaryCategorySchemaList = yup
  .array()
  .of(primaryCategorySchema)
  .required();

export const pagedPrimaryCategoriesSchema = pagedResultsSchema.shape({
  results: primaryCategorySchemaList,
});

export type Category = yup.InferType<typeof categorySchema>;

export type CategoryInput = yup.InferType<typeof categoryInputSchema>;

export type Subcategory = yup.InferType<typeof subcategorySchema>;

export type PrimaryCategory = yup.InferType<typeof primaryCategorySchema>;

export const CategoryInputLabels: Labels<CategoryInput> = {
  id: "ID",
  name: "Name",
  parentCategory: "Parent Category",
  owner: "Owner",
};

export const findCatById = (id: string, cats: Category[]) => {
  let result: Category = {
    id: UNCATEGORIZED,
    name: "Uncategorized",
    owner: SYSTEM_USER,
  };
  let findResult = cats.find((cat) => cat.id === id);
  if (findResult) {
    result = findResult;
  }
  return result;
};

export function sortCategories<C extends Category>(
  categories: C[],
  uncategorizedPos: "first" | "last" | "alpha" = "last"
) {
  return categories.sort((catA, catB) => {
    if (catA.id === UNCATEGORIZED && uncategorizedPos !== "alpha") {
      return uncategorizedPos === "last" ? 1 : -1;
    } else if (catB.id === UNCATEGORIZED && uncategorizedPos !== "alpha") {
      return uncategorizedPos === "last" ? -1 : 1;
    }
    return catA.name.localeCompare(catB.name);
  });
}
